lib/model/compose.dart

```diff
@@ -0,0 +1,103 @@
+  const lineStart = r'^ {0,3}';
+
+  // The backticks, captured so we can see how many.
+  const backticks = r'(`{3,})';
+
+  // The "info string" plus (meaningless) leading or trailing spaces or tabs.
+  // It can't contain backticks.
+  const trailing = r'[^`]*$';
+  return RegExp(lineStart + backticks + trailing, multiLine: true);
+})();
+
+/// The shortest opening backtick fence that's longer than any in [content].
+///
+/// Expressed as a number of backticks.
+///
+/// Use this for quote-and-reply or anything else that requires wrapping
+/// Markdown in a backtick fence.
+///
+/// See the CommonMark spec, which Zulip servers should but don't always follow:
+///   https://spec.commonmark.org/0.30/#fenced-code-blocks
+int getUnusedOpeningBacktickFenceLength(String content) {
+  final matches = _openingBacktickFenceRegex.allMatches(content);
+  int result = 3;
+  for (final match in matches) {
+    result = max(result, match[1]!.length + 1);
+  }
+  return result;
+}
+
+/// Wrap Markdown [content] with opening and closing backtick fences.
+///
+/// For example, for this Markdown:
+///
+/// ```javascript
+/// console.log('Hello world!');
+/// ```
+///
+/// this function, with `infoString: 'quote'`, gives
+///
+/// ````quote
+/// ```javascript
+/// console.log('Hello world!');
+/// ```
+/// ````
+///
+/// See the CommonMark spec, which Zulip servers should but don't always follow:
+///   https://spec.commonmark.org/0.30/#fenced-code-blocks
+// In [content], indented code blocks
+//   ( https://spec.commonmark.org/0.30/#indented-code-blocks )
+// and code blocks fenced with tildes should make no difference to the
+// backtick fences we choose here; this function ignores them.
+String wrapWithBacktickFence({required String content, String? infoString}) {
+  assert(infoString == null || !infoString.contains('`'));
+  assert(infoString == null || infoString.trim() == infoString);
+
+  StringBuffer resultBuffer = StringBuffer();
+
+  // (A) Why not panic on dangling opening fences
+  //     (ones without a corresponding closing fence)?
+  //     Because the spec allows leaving the closing fence implicit:
+  //     > If the end of the containing block (or document) is reached
+  //     > and no closing code fence has been found,
+  //     > the code block contains all of the lines after the opening code fence
+  //     > until the end of the containing block (or document).
+  //
+  // (B) Why not look for dangling closing fences (ones without an opening fence)?
+  //     Because technically there's no such thing:
+  //     they would be indistinguishable from dangling opening fences,
+  //     and parsers (and this function) will treat them that way.
+  final fenceLength = getUnusedOpeningBacktickFenceLength(content);
+
+  resultBuffer.write('`' * fenceLength);
+  if (infoString != null) {
+    resultBuffer.write(infoString);
+  }
+  resultBuffer.write('\n');
+  resultBuffer.write(content);
+  resultBuffer.write('\n');
```

Comment:
So one thing this function is doing is it's effectively assuming that its argument `content` is missing its final newline — that it ends with an incomplete line. If the argument is instead made of complete lines, then this will add a blank line at the end before the closing fence.

If you look back at the spec: https://spec.commonmark.org/0.30/#fenced-code-blocks
and at the HTML output for the examples, the content (enclosed in `<code>…</code>`) always consists of complete lines — it's always either empty (so consists of zero lines, all of them complete), or ends with a newline. The only exceptions are examples 121, 138, and 145… which demonstrate syntax that does *not* make a fenced code block, and instead makes some inline code.

And that's reflected in the spec text: "The content of the code block consists of all subsequent _lines_", emphasis added.

I think the norm for this function should therefore be that the input consists of complete lines, just as the output will. Probably it's best if it also cheerfully accepts an incomplete final line, and supplies its newline. The implementation then just needs to say:
```suggestion
  resultBuffer.write(content);
  if (!content.endsWith('\n')) {
    resultBuffer.write('\n');
  }
```